src/neosophia/agents/prompt.py

```diff
@@ -1,4 +1,46 @@
 """ Class for generating a structured prompt """
+import re
+
+from functools import partial
+
+import pandas as pd
+
+
+def format_df(df):
+    """
+    Function that stripts out all whitespace between dataframe columns while
+    preserving white space in individual cells
+    """
+    def get_fmt_str(x, fill):
+        return '| {message: >{fill}} '.format(message=x, fill=fill-2)
+
+    # Max character length per column
+    s = df.astype(str).agg(lambda x: x.str.len()).max()
+    pad = 0  # How many spaces between
+    fmts = {}
+    header_strs = []
+    for idx, c_len in s.items():
+        if isinstance(idx, tuple):
+            lab_len = max([len(str(x)) for x in idx])
+        else:
+            lab_len = len(str(idx))
+
+        fill = max(lab_len, c_len) + pad
+        fmts[idx] = partial(get_fmt_str, fill=fill)
+
+        # Formatting the header
+        header_strs.append(get_fmt_str(idx, fill))
```

Comment:
If I'm reading this correctly, it's calculating (incorrectly) how many spaces around the pipes are required to add to make a nicely formatted dataframe, which you are then removing with the regex at the end 😆

Reply:
This appears to be unchanged